### ISSUE TITLE
[css-values] Consider requiring same relative <length> units in atan2()

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#8169]
Consider `atan2(10em, 13vh)`. Should that work? Per spec, it should because the two units are `<length>`s, and [the spec](https://drafts.csswg.org/css-values-4/#trig-funcs) says:

> The atan2(A, B) function contains two comma-separated [calculations](https://drafts.csswg.org/css-values-4/#calc-calculation), A and B. A and B can resolve to any [`<number>`](https://drafts.csswg.org/css-values-4/#number-value), [`<dimension>`](https://drafts.csswg.org/css-values-4/#typedef-dimension), or [`<percentage>`](https://drafts.csswg.org/css-values-4/#percentage-value), but must have the same [type](https://drafts.csswg.org/css-values-4/#determine-the-type-of-a-calculation), or else the function is invalid.

If we allow different units, we can't resolve `atan()` at parse-time (because it starts depending on resolving those `<length>`s to pixels). That introduces a new set of property dependencies which didn't exist before. For example, we'd need to define how ex/etc units resolve on `font-weight`, since `font-weight` can affect the primary font, and with this you could do `font-weight: atan2(2ex, 10px)`.

My implementation in Firefox doesn't allow mixed relative units in `atan2()` (because I misread the spec). This is also problematic for Chromium implementation-wise (even without all these new dependency shenanigans).

Unless this is extremely useful, it might be worth restricting relative lengths to enforce same-unit-ness.

cc: @tabatkins @devsdk @dbaron

**Comment by tabatkins** [comment]
Yeah, I think this *is* pretty important - for example, `atan2(10vw, 10vh)` makes complete sense to me.

But more importantly, this isn't an atan2()-specific issue; it applies to *everything* taking a calculation. `font-weight: calc(2ex / 10px)`, for example, introduces the exact same problem. So I'm strongly against any atan2()-targeted change here; if we want to disallow this we need to instead apply some more powerful restriction to all calculations.

**Comment by Loirooriol** [comment]
Not sure if I follow, is the "primary font" the same as the "first available font"? Then, what about `font-weight: 2ex` without calculations, the `2ex` resolves depending on the "first available font", which depends on `font-weight`?

**Comment by xiaochengh** [comment]
We can't plug relative length units into `font-weight` without calculations, because `font-weight` takes number values only.

I think we can just use the same idea how we resolve `font-size: 1ex`: when a font-relative length unit is used in a font-related property, it refers to the the parent element's font.

**Comment by Loirooriol** [comment]
Ah, sure, time to go to sleep I guess XD. But I think this would also need to address custom properties registered as `<integer>` or `<number>` (https://drafts.css-houdini.org/css-properties-values-api/#dependency-cycles):
```css
--foo: calc(2ex / 10px);
font-weight: var(--foo);
```

**Comment by Loirooriol** [comment]
Also:

```css
line-height: 1ex;
font-weight: calc(1lh / 1px);
```

So, generalizing #2115, I guess the idea would be:

1. Compute `font-size`, resolving unknown font units with the parent. This sets the `em` unit.
2. Compute `font-weight`, resolving unknown font units with the parent. This sets the `ex`, `cap`, `ch` and `ic` units.
3. Compute `line-height`, resolving unknown font units with the parent. This sets the `lh` unit.

Or:

1. Compute `font-weight`, resolving unknown font units with the parent.
2. Compute `font-size`, resolving unknown font units with the parent. This sets the `em`, `ex`, `cap`, `ch` and `ic` units.
3. Compute `line-height`, resolving unknown font units with the parent. This sets the `lh` unit.

**Comment by Loirooriol** [comment]
Retagging to values-4 since values-3 doesn't have `atan2` nor unit algebra.

**Comment by emilio** [comment]
font-style needs to also be special cased along with font-weight at least, since you would be allowed to set a slant anglebased on any unit. Probably others as well.

**Comment by Loirooriol** [comment]
`font-stretch` too I guess. https://drafts.csswg.org/css-fonts-4/#font-prop-desc

**Comment by Loirooriol** [comment]
All of this seems problematic:
```css
font-weight: calc(1ex / 1px);
font-style: calc(1ex / 1px * 1deg);
font-stretch: calc(1ex / 1px * 1%);
font-variation-settings: "wght" calc(1ex / 1px);
font-feature-settings: "subs" calc(1ex / 1px);
font-size-adjust: ex-height calc(1ex / 1px);
```

**Comment by yukulele** [comment]
> Unless this is extremely useful, it might be worth restricting relative lengths to enforce same-unit-ness.

In this case the units would be useless because `atan2(X, Y)` = `atan2(Xunit, Yunit)`:
```css
rotate: atan2(12, 7)     /* = 1.04272rad */
rotate: atan2(12%, 7%)   /* = 1.04272rad */
rotate: atan2(12em, 7em) /* = 1.04272rad */
rotate: atan2(12vw, 7vw) /* = 1.04272rad */
rotate: atan2(12ms, 7ms) /* = 1.04272rad */
...
```

**Comment by yukulele** [comment]
https://github.com/w3c/csswg-drafts/issues/7482

**Comment by yukulele** [comment]
possible duplicate of [7482](https://github.com/w3c/csswg-drafts/issues/7482)

**Comment by tabatkins** [comment]
Nah, #7482 was a question about what the behavior was. This is an issue asking if we should *change* the behavior.

**Comment by xiaochengh** [comment]
Related: what if we use them in animation properties? For example

```css
.test {
  animation-duration: calc(sin(atan2(1em, 20px)) * 1s);
  animation-name: anim;
}

@keyframes anim {
  from { font-size: 10px; }
  to { font-size: 30px; }
}
```

Do they need the same "[animation-tainted](https://www.w3.org/TR/css-variables-1/#animation-tainted)" treatment as in custom properties?

**Comment by emilio** [comment]
`sign()` has the same issue.

**Comment by tabatkins** [comment]
Yes, as I said in [my earlier comment](https://github.com/w3c/csswg-drafts/issues/8169#issuecomment-1334573630), it applies to *every* function taking a calculation, including `calc()` itself.

**Comment by DevSDK** [comment]
@Loirooriol 

I'm trying to understand given example why has circular dependency:
```
line-height: 1ex;
font-weight: calc(1lh / 1px);
```

Can I ask the spec references that the font-weight affects x-height?

**Comment by emilio** [comment]
@DevSDK any font which affects the primary font selection can change the first available font, which at least in theory can affects font metrics, including x-height...

**Comment by Loirooriol** [comment]
> ACTION TabAtkins: Figure out what to do on custom properties typed as lengths

The problem is not just for custom properties registered as lengths, see https://github.com/w3c/css-houdini-drafts/issues/1080

**Comment by Loirooriol** [comment]
Mild concern with the resolution: this creates an inconsistency with #2115, i.e.

```css
font-size: 2lh; /* resolves using the parent */
line-height: 3em; /* resolves using the current element */
```

```css
font-size: 2ex; /* resolves using the parent */
font-weight: calc(3em / 1px); /* resolves using the parent */
```

Since `font-size` always resolves font-relative units with the parent, it should be possible to resolve `em` with the current element both for `line-height` and `font-size`, so why do different things?

**Comment by fantasai** [comment]
@Loirooriol (I assume you mean `font-weigh` in your last sentence.) It would be really bad if, in `font-weight`, `3em` resolved against the element itself but `3ex` resolved against the parent. Since `font-weight` can affect the calculation of `ex`, `ex` has to resolve against the parent.

**Comment by Loirooriol** [comment]
@fantasai Why would resolving `em` and `ex` against different elements be really bad, but resolving `em` and `lh` against different elements is fine? Should all units resolve against the parent when there is a cycle?

**Comment by tabatkins** [comment]
We can't change the behavior of `line-height: 1em;` in general (it uses current-element ems), and special-casing it to change behavior if and only if `font-size` was in `lh` is a difficult-to-predict, action-at-a-distance sort of thing that we generally want to avoid. But we have to break the cycle *somehow*, so having `font-size: 1lh;` always resolve against the parent was the most reasonable thing to do.

Digging in a little deeper, tho, while line-height and font-size usually establish a dependency, they're not *semantically connected* - the font size and the height of a line don't readily appear to be different aspect of the same concept, at least imo. It's thus more acceptable to have them differ on which element they refer to, in these cases.

On the other hand, `em` and `ex` (and all the other explicitly font-based units) *are* conceptually related; they're all sizing according to some text, and it's not uncommon for them to be used in parallel fashions. Having them differ on what element they refer to would make the grouping conceptually incoherent.

**Comment by yisibl** [comment]
The ability to convert between different units in atan2() is very useful, E.g. we can now do a pure CSS real-time display of the container width(px).

Demo: https://codepen.io/xboxyan/pen/MWZLYZz

https://github.com/w3c/csswg-drafts/assets/2784308/de0d47b3-716a-4a92-8082-d0ebc2d3f46b
